src/aggregation/metric/top_hits.rs

```diff
@@ -0,0 +1,820 @@
+}
+
+const fn default_doc_value_fields() -> Vec<String> {
+    Vec::new()
+}
+
+/// Search query spec for each matched document
+/// TODO: move this to a common module
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
+pub struct RetrievalFields {
+    /// The fast fields to return for each hit.
+    /// This is the only variant supported for now.
+    /// TODO: support the {field, format} variant for custom formatting.
+    #[serde(rename = "docvalue_fields")]
+    #[serde(default = "default_doc_value_fields")]
+    pub doc_value_fields: Vec<String>,
+}
+
+/// Search query result for each matched document
+/// TODO: move this to a common module
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
+pub struct FieldRetrivalResult {
+    /// The fast fields returned for each hit.
+    #[serde(rename = "docvalue_fields")]
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    pub doc_value_fields: HashMap<String, OwnedValue>,
+}
+
+impl RetrievalFields {
+    fn get_field_names(&self) -> Vec<&str> {
+        self.doc_value_fields.iter().map(|s| s.as_str()).collect()
+    }
+
+    fn resolve_field_names(self, reader: &ColumnarReader) -> crate::Result<Self> {
+        let globbed_string_to_regex = |globbed: &str| {
+            let sanitized = format!("^{}$", regex::escape(globbed).replace(r"\*", ".*"));
+            Regex::new(&sanitized.replace('*', ".*")).map_err(|e| {
+                crate::TantivyError::SchemaError(format!(
+                    "Invalid regex '{}' in docvalue_fields: {}",
+                    globbed, e
+                ))
+            })
+        };
+        let resolved_fields = self
+            .doc_value_fields
+            .into_iter()
+            .map(|field| {
+                if !field.contains('*') && reader.iter_columns()?.any(|(name, _)| name == field) {
+                    return Ok(vec![field]);
+                }
+
+                let pattern = globbed_string_to_regex(&field)?;
+                Ok(reader
+                    .iter_columns()?
+                    .filter(|(name, _)| pattern.is_match(name))
+                    .map(|(name, _)| name)
+                    .collect::<Vec<_>>())
+            })
+            .collect::<crate::Result<Vec<_>>>()?
+            .into_iter()
+            .flatten()
+            .collect();
+
+        Ok(Self {
+            doc_value_fields: resolved_fields,
+        })
+    }
+
+    fn get_fields(
+        &self,
+        accessors: &HashMap<String, DynamicColumn>,
+        doc_id: DocId,
+    ) -> FieldRetrivalResult {
+        let dvf = self
+            .doc_value_fields
+            .iter()
+            .map(|field| {
+                let accessor = accessors.get(field).expect("field accessor must exist");
+
+                let value: OwnedValue = match accessor {
+                    DynamicColumn::U64(accessor) => accessor
+                        .values_for_doc(doc_id)
+                        .next()
```

Comment:
We should fetch all the data, not just the first one

Reply:
[d760c6c](https://github.com/quickwit-oss/tantivy/pull/2198/commits/d760c6c840b050f1ca3053af5b6f68c2f4afadcb)